### PR TITLE
Fix loss explosion on resume when training on small dataset

### PR DESCRIPTION
**Pull request by Y-T-G**
Closes #24079 

## Claude summary

In 8.4.8, the optimizer selection changed from always using MuSGD to using AdamW for small datasets (<10k iterations). Unlike SGD, AdamW maintains exp_avg_sq (second moment estimates) which contain very small values near zero. When convert_optimizer_state_dict_to_fp16 converts these to FP16 during checkpoint saving, values below ~6e-5 are rounded to zero. On resume,Adam's update rule divides by sqrt(exp_avg_sq) + eps, and the zeroed values cause divide-by-epsilon explosions, spiking cls_loss by 1000-5000x.  

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
🛠️ This PR refines optimizer state conversion to FP16 by keeping the `exp_avg_sq` tensor in FP32, improving numerical stability during training.

### 📊 Key Changes
- Updated `convert_optimizer_state_dict_to_fp16()` in `ultralytics/utils/torch_utils.py`.
- Changed the FP16 conversion rule so that optimizer state entries named `exp_avg_sq` are **no longer converted** from `float32` to `float16`.
- The function still skips converting `step`, and now also skips `exp_avg_sq`.

### 🎯 Purpose & Impact
- ✅ **Improves training stability** by preserving a sensitive optimizer statistic in higher precision.
- ⚙️ **Reduces risk of precision-related issues** when resuming training or working with mixed-precision workflows.
- 🚀 **Helps maintain reliable optimizer behavior** without broadly changing the FP16 conversion logic.
- 👥 **Benefits users training YOLO models at scale**, especially in setups where checkpoint efficiency and stable optimization both matter.

### CODE REVIEW

File: ultralytics/utils/torch_utils.py [hunk 1]
```diff
@@ -770,7 +770,7 @@ def convert_optimizer_state_dict_to_fp16(state_dict):
     """
     for state in state_dict["state"].values():
         for k, v in state.items():
-            if k != "step" and isinstance(v, torch.Tensor) and v.dtype is torch.float32:
+            if k not in {"step", "exp_avg_sq"} and isinstance(v, torch.Tensor) and v.dtype is torch.float32:
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/torch_utils.py:773]: 💡 **MEDIUM**: This fixes Adam/AdamW's `exp_avg_sq`, but AMSGrad keeps a second-moment cap in `max_exp_avg_sq` that has the same underflow risk when cast to FP16. Resuming an AMSGrad checkpoint can still zero out that denominator state and produce the same instability.

**Suggested change:**
```suggestion
            if k not in {"step", "exp_avg_sq", "max_exp_avg_sq"} and isinstance(v, torch.Tensor) and v.dtype is torch.float32:
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The optimizer-state fix looks well targeted and the docs change is harmless. One follow-up worth addressing: AMSGrad's `max_exp_avg_sq` should be preserved in FP32 too, otherwise the same resume-time instability can still occur for that optimizer variant.

💬 Posted 1 inline comment

File: ultralytics/utils/torch_utils.py (hunk 1 above)
Comment (Y-T-G) [reply ultralytics/utils/torch_utils.py:773]: We don't have AMSGrad

**Comment by Laughing-q** [comment]
@Y-T-G the tests we talked about before still raises the same cls loss increasing.
```bash
yolo train model=yolo26m.pt data=coco128-seg.yaml batch=2 nbs=1 name=resume_test exist_ok
# interrupt at the 3rd epoch then run
yolo train model=runs/detect/resume_test/weights/last.pt device=-1 resume=True
```

**Comment by Y-T-G** [comment]
@Laughing-q I am using the segmentation model to test now:

**main**
Interrupt after 8 epochs
`yolo train model=yolo26m-seg.pt device=-1 data=coco128-seg.yaml batch=2 nbs=1 name=resume_test exist_ok`
```

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss   sem_loss  Instances       Size
      1/100      2.07G      1.129      2.429       1.87   0.009216     0.3936          7        640: 100% ━━━━━━━━━━━━ 64/64 8.0it/s 8.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 32/32 3.4it/s 9.3s
                   all        128        929      0.836      0.612      0.741      0.575      0.813      0.597      0.697       0.48

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss   sem_loss  Instances       Size
      2/100      2.44G      1.216      2.227      1.476   0.008495     0.3063         27        640: 100% ━━━━━━━━━━━━ 64/64 11.5it/s 5.6s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 32/32 34.3it/s 0.9s
                   all        128        929      0.699      0.709      0.786      0.586      0.682      0.688      0.737      0.493

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss   sem_loss  Instances       Size
      3/100      2.53G      1.308      2.406      1.611    0.01063     0.3156         38        640: 100% ━━━━━━━━━━━━ 64/64 15.3it/s 4.2s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 32/32 34.9it/s 0.9s
                   all        128        929      0.766      0.595      0.728      0.503      0.727       0.56      0.659      0.417

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss   sem_loss  Instances       Size
      4/100      2.53G      1.433      2.579      1.898    0.01137     0.3189         34        640: 100% ━━━━━━━━━━━━ 64/64 15.2it/s 4.2s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 32/32 36.4it/s 0.9s
                   all        128        929      0.738      0.589      0.685      0.465      0.693      0.549      0.614      0.372

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss   sem_loss  Instances       Size
      5/100      2.53G       1.36      2.449      1.654    0.01135     0.2943          9        640: 100% ━━━━━━━━━━━━ 64/64 14.5it/s 4.4s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 32/32 37.0it/s 0.9s
                   all        128        929      0.719      0.634      0.736       0.52      0.684      0.609      0.684      0.426

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss   sem_loss  Instances       Size
      6/100      2.53G        1.4      2.426      1.658    0.01122     0.2858         22        640: 100% ━━━━━━━━━━━━ 64/64 14.3it/s 4.5s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 32/32 37.2it/s 0.9s
                   all        128        929      0.747      0.599      0.726       0.51      0.712      0.562      0.674      0.431

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss   sem_loss  Instances       Size
      7/100      2.53G      1.332      2.457      1.648     0.0112     0.2845         20        640: 100% ━━━━━━━━━━━━ 64/64 14.7it/s 4.4s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 32/32 34.4it/s 0.9s
                   all        128        929      0.744      0.672      0.781      0.555      0.728      0.617      0.702      0.429

```

Resume
`yolo train model=runs/segment/resume_test/weights/last.pt device=-1 data=coco128-seg.yaml batch=2 nbs=1 resume=True`
```
      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss   sem_loss  Instances       Size
      8/100      2.16G      2.094       4.68  2.029e+04    0.02495     0.7366          7        640: 100% ━━━━━━━━━━━━ 64/64 7.7it/s 8.3s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 32/32 3.4it/s 9.3s
                   all        128        929          0          0          0          0          0          0          0          0

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss   sem_loss  Instances       Size
      9/100      2.68G      1.985       3.24       5683    0.02006     0.5336         27        640: 100% ━━━━━━━━━━━━ 64/64 14.1it/s 4.5s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 32/32 34.9it/s 0.9s
                   all        128        929     0.0775     0.0129     0.0458      0.021     0.0563    0.00832     0.0326      0.011
```

**PR**

Train
```
      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss   sem_loss  Instances       Size
      1/100      2.07G      1.129      2.429       1.87   0.009216     0.3936          7        640: 100% ━━━━━━━━━━━━ 64/64 7.6it/s 8.4s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 32/32 3.5it/s 9.2s
                   all        128        929      0.836      0.612      0.741      0.575      0.813      0.597      0.697       0.48

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss   sem_loss  Instances       Size
      2/100      2.51G      1.216      2.227      1.476   0.008495     0.3063         27        640: 100% ━━━━━━━━━━━━ 64/64 15.0it/s 4.3s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 32/32 33.6it/s 1.0s
                   all        128        929      0.699      0.709      0.786      0.586      0.682      0.688      0.737      0.493

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss   sem_loss  Instances       Size
      3/100       2.6G      1.308      2.406      1.611    0.01063     0.3156         38        640: 100% ━━━━━━━━━━━━ 64/64 10.9it/s 5.9s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 32/32 36.0it/s 0.9s
                   all        128        929      0.766      0.595      0.728      0.503      0.727       0.56      0.659      0.417

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss   sem_loss  Instances       Size
      4/100       2.6G      1.433      2.579      1.898    0.01137     0.3189         34        640: 100% ━━━━━━━━━━━━ 64/64 14.6it/s 4.4s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 32/32 36.1it/s 0.9s
                   all        128        929      0.738      0.589      0.685      0.465      0.693      0.549      0.614      0.372

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss   sem_loss  Instances       Size
      5/100       2.6G       1.36      2.449      1.654    0.01135     0.2943          9        640: 100% ━━━━━━━━━━━━ 64/64 14.9it/s 4.3s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 32/32 34.3it/s 0.9s
                   all        128        929      0.719      0.634      0.736       0.52      0.684      0.609      0.684      0.426

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss   sem_loss  Instances       Size
      6/100       2.6G        1.4      2.426      1.658    0.01122     0.2858         22        640: 100% ━━━━━━━━━━━━ 64/64 15.2it/s 4.2s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 32/32 34.4it/s 0.9s
                   all        128        929      0.747      0.599      0.726       0.51      0.712      0.562      0.674      0.431

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss   sem_loss  Instances       Size
      7/100       2.6G      1.332      2.457      1.648     0.0112     0.2845         20        640: 100% ━━━━━━━━━━━━ 64/64 14.9it/s 4.3s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 32/32 34.7it/s 0.9s
                   all        128        929      0.744      0.672      0.781      0.555      0.728      0.617      0.702      0.429
```

Resume
```
      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss   sem_loss  Instances       Size
      8/100      2.16G      1.295      2.322      1.713    0.01222     0.2975          7        640: 100% ━━━━━━━━━━━━ 64/64 7.1it/s 9.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 32/32 3.4it/s 9.3s
                   all        128        929      0.711      0.694      0.779      0.545      0.664      0.642      0.703      0.443

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss   sem_loss  Instances       Size
      9/100      2.54G      1.289       2.12      1.333    0.01031     0.2391         27        640: 100% ━━━━━━━━━━━━ 64/64 14.9it/s 4.3s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 32/32 37.1it/s 0.9s
                   all        128        929      0.762      0.765      0.835      0.611      0.729      0.738       0.77      0.507

```

**Review by Laughing-q — APPROVED** [review]

**Comment by Laughing-q** [comment]
@Y-T-G oh yeah I think you are right! I was testing using the last.pt trained from main branch so actually the fix here is not included in that pt file. I just re-tested with training and resume from this branch, all good!

**Comment by Laughing-q** [comment]
@Y-T-G Awesome work on this! merging!

**Comment by UltralyticsAssistant** [comment]
🎉 Thrilled to see this merged — huge thanks to @Y-T-G for the thoughtful improvement, and to @Laughing-q for the valuable contributions along the way!

Keeping `exp_avg_sq` in FP32 is a small but impactful refinement that strengthens training stability and makes mixed-precision and checkpoint workflows more reliable for the broader YOLO community. As Leonardo da Vinci said, *“Details make perfection, and perfection is not a detail.”* This PR is a great example of how careful attention to the right detail can meaningfully improve the training experience at scale.

Really appreciate the care and precision behind this contribution — thank you both! 🙌
